src/data_gradients/managers/abstract_manager.py

```diff
@@ -95,38 +97,88 @@ def execute(self):
             "computer vision datasets. click here: https://hubs.ly/Q01XpHBT0"
         )
 
-        datasets_tqdm = tqdm(
-            zip_longest(self.train_samples_iterator, self.val_samples_iterator, fillvalue=None),
-            desc="Analyzing... ",
-            total=self.n_batches,
-        )
+        # datasets_tqdm = iter(tqdm(
+        #     zip_longest(self.train_samples_iterator, self.val_samples_iterator, fillvalue=None),
+        #     desc="Analyzing... ",
+        #     total=self.n_batches,
+        # ))
+
+        train_samples_iterator = iter(self.train_samples_iterator)
+        val_samples_iterator = iter(self.val_samples_iterator)
 
         self._train_iters_done, self._val_iters_done = 0, 0
         self._stopped_early = False
 
-        for i, (train_sample, val_sample) in enumerate(datasets_tqdm):
+        non_finite_train_samples = 0
+        non_finite_train_samples_example = ""
+        non_finite_val_samples = 0
+        non_finite_val_samples_example = ""
+
+        with tqdm(total=self.n_batches, desc="Analyzing... ") as datasets_tqdm:
+            i = 0
+            while True:
+                if i == self.batches_early_stop:
+                    self._stopped_early = True
+                    break
 
-            if i == self.batches_early_stop:
-                self._stopped_early = True
-                break
+                try:
+                    train_sample = None
```

Comment:
IS that required? wouldn't `next` return None if the iterator ended. 
If it is just an init, can you please move this line outside of the loop?

Reply:
> wouldn't next return None if the iterator ended.
Nope, when iterator is exhausted it raises StopIteration exception and you should wait for it to know when it's ended.
So there are a few stopping criterias:
1) `batches_early_stop`
2) Both train and validation iterators are exhausted 

So yes, it is needed

Reply:
In e8f3e923454842e4a4f6c5446984f6c4100b2e67 I've moved exit criterion checks to a single location at the end of the loop (easier to follow when all breaks are in a single place)